File: core/src/commonMain/kotlin/com/xhsdn/core/parse/MediaUrlExtractor.kt
package com.xhsdn.core.parse

import kotlinx.serialization.json.JsonArray
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.jsonArray
import kotlinx.serialization.json.jsonObject
import kotlinx.serialization.json.jsonPrimitive

/**
 * 从 note JSON 中提取媒体 URL 与 Live Photo 配对。
 * 复刻 [XHSDownloader.extractMediaUrlsFromNote]（java:1143-1331）的核心逻辑。
 */
object MediaUrlExtractor {

    /**
     * 单张图 / 视频配对。`isLivePhoto == true` 时 [videoUrl] 不为空。
     */
    data class MediaPair(
        var originalImageUrl: String?,
        var originalVideoUrl: String?,
        var imageUrl: String?,
        var videoUrl: String?,
        var isLivePhoto: Boolean,
    )

    /** 视频回调：用于触发 "检测到视频" 通知。 */
    fun interface OnVideoDetected {
        fun onVideoDetected()
    }

    /**
     * 抽取媒体 URL 与 Live Photo 配对。
     * @return 媒体 URL 列表（图/视频混合）
     */
    fun extract(
        note: JsonObject,
        mediaPairs: MutableList<MediaPair>,
        onVideoDetected: OnVideoDetected? = null,
    ): List<String> {
        val mediaUrls = mutableListOf<String>()
        var videosDetected = false

        // 1. 视频主帖
        val video = note["video"] as? JsonObject
        if (video != null) {
            val consumer = video["consumer"] as? JsonObject
            val originKey = consumer?.get("originVideoKey")?.jsonPrimitiveOrNull()
            if (originKey != null) {
                val videoUrl = "https://sns-video-bd.xhscdn.com/$originKey"
                mediaUrls += videoUrl
                videosDetected = true
                onVideoDetected?.onVideoDetected()
            } else {
                val media = video["media"] as? JsonObject
                val stream = media?.get("stream") as? JsonObject
                val h265 = stream?.get("h265") as? JsonArray
                if (h265 != null) {
                    for (item in h265) {
                        val url = when (item) {
                            is kotlinx.serialization.json.JsonPrimitive -> {
                                val s = item.content
                                if (s.startsWith("http")) s else null
                            }
                            is JsonObject -> {
                                item["url"]?.let { it.jsonPrimitiveOrNull() }
                                    ?: item["masterUrl"]?.let { it.jsonPrimitiveOrNull() }
                            }
                            else -> null
                        }
                        if (url != null) {
                            mediaUrls += url
                            videosDetected = true
                            onVideoDetected?.onVideoDetected()
                        }
                    }
                }
            }
        }

        // 2. 图片列表（图集 / Live Photo）
        val imageList: JsonArray? = when {
            note["imageList"] is JsonArray -> (note["imageList"] as JsonArray)
            note["images"] is JsonArray -> (note["images"] as JsonArray)
            note["image"] is JsonObject -> {
                val single = note["image"] as JsonObject
                JsonArray(listOf(single))
            }
            else -> null
        }

        if (imageList != null) {
            for (element in imageList) {
                val image = element as? JsonObject ?: continue
                val imageUrl = pickString(image, listOf("urlDefault", "url"))
                    ?: run {
                        val traceId = image["traceId"]?.let { it.jsonPrimitiveOrNull() }
                        if (traceId != null) "https://sns-img-qc.xhscdn.com/$traceId" else null
                    }
                    ?: run {
                        val infoList = image["infoList"] as? JsonArray
                        infoList?.firstOrNull()?.let { (it as? JsonObject)?.get("url")?.let { u -> u.jsonPrimitiveOrNull() } }
                    }

                // Live Photo: image.stream.h264[0].masterUrl / url
                val livePhotoVideoUrl = run {
                    val stream = image["stream"] as? JsonObject
                    val h264 = stream?.get("h264") as? JsonArray
                    val first = h264?.firstOrNull()
                    when (first) {
                        is JsonObject -> {
                            first["masterUrl"]?.let { it.jsonPrimitiveOrNull() }
                                ?: first["url"]?.let { it.jsonPrimitiveOrNull() }
                        }
                        else -> null
                    }
                }

                if (imageUrl != null) {
                    if (livePhotoVideoUrl != null) {
                        mediaPairs += MediaPair(imageUrl, livePhotoVideoUrl, imageUrl, livePhotoVideoUrl, true)
                        videosDetected = true
                        onVideoDetected?.onVideoDetected()
                    } else {
                        mediaPairs += MediaPair(imageUrl, null, imageUrl, null, false)
                    }
                }
            }
        }

        return mediaUrls
    }

    private fun pickString(obj: JsonObject, keys: List<String>): String? {
        for (k in keys) {
            val v = obj[k]?.let { it.jsonPrimitiveOrNull() }
            if (v != null) return v
        }
        return null
    }

    private fun kotlinx.serialization.json.JsonElement.jsonPrimitiveOrNull(): String? =
        (this as? kotlinx.serialization.json.JsonPrimitive)?.let { if (it.isString) it.content else null }
}

File: web/src/core/parse/MediaUrlExtractor.ts
import { transformXhsCdnUrl } from '../url/UrlTransformer';

/**
 * 从 note JSON 中提取媒体 URL 与 Live Photo 配对。
 *
 * 1:1 翻译自 `core/src/commonMain/kotlin/com/xhsdn/core/parse/MediaUrlExtractor.kt`。
 */

type Json = null | boolean | number | string | Json[] | { [k: string]: Json };
type JsonObject = { [k: string]: Json };
type JsonArray = Json[];

export interface MediaPair {
  originalImageUrl: string | null;
  originalVideoUrl: string | null;
  imageUrl: string | null;
  videoUrl: string | null;
  isLivePhoto: boolean;
}

export type OnVideoDetected = () => void;

function isObject(v: Json | undefined): v is JsonObject {
  return !!v && typeof v === 'object' && !Array.isArray(v);
}
function isArray(v: Json | undefined): v is JsonArray {
  return Array.isArray(v);
}
function asString(v: Json | undefined): string | null {
  return typeof v === 'string' ? v : null;
}

function pickString(obj: JsonObject, keys: string[]): string | null {
  for (const k of keys) {
    const v = asString(obj[k]);
    if (v !== null) return v;
  }
  return null;
}

export function extract(
  note: JsonObject,
  mediaPairs: MediaPair[],
  onVideoDetected?: OnVideoDetected,
): string[] {
  const mediaUrls: string[] = [];
  let videosDetected = false;

  // 1. 视频主帖
  const video = note['video'];
  if (isObject(video)) {
    const consumer = video['consumer'];
    if (isObject(consumer)) {
      const originKey = asString(consumer['originVideoKey']);
      if (originKey) {
        const videoUrl = `https://sns-video-bd.xhscdn.com/${originKey}`;
        mediaUrls.push(videoUrl);
        videosDetected = true;
        if (onVideoDetected) onVideoDetected();
      }
    } else {
      const media = video['media'];
      if (isObject(media)) {
        const stream = media['stream'];
        if (isObject(stream)) {
          const h265 = stream['h265'];
          if (isArray(h265)) {
            for (const item of h265) {
              let url: string | null = null;
              if (typeof item === 'string') {
                url = item.startsWith('http') ? item : null;
              } else if (isObject(item)) {
                url = asString(item['url']) ?? asString(item['masterUrl']);
              }
              if (url) {
                mediaUrls.push(url);
                videosDetected = true;
                if (onVideoDetected) onVideoDetected();
              }
            }
          }
        }
      }
    }
  }

  // 2. 图片列表(图集 / Live Photo)
  let imageList: JsonArray | null = null;
  if (isArray(note['imageList'])) {
    imageList = note['imageList'] as JsonArray;
  } else if (isArray(note['images'])) {
    imageList = note['images'] as JsonArray;
  } else if (isObject(note['image'])) {
    imageList = [note['image'] as JsonObject];
  }

  if (imageList) {
    for (const element of imageList) {
      if (!isObject(element)) continue;
      const image = element;
      const imageUrl =
        pickString(image, ['urlDefault', 'url']) ??
        (() => {
          const traceId = asString(image['traceId']);
          return traceId ? `https://sns-img-qc.xhscdn.com/${traceId}` : null;
        })() ??
        (() => {
          const infoList = image['infoList'];
          if (isArray(infoList)) {
            const first = infoList[0];
            if (isObject(first)) return asString(first['url']);
          }
          return null;
        })();

      // Live Photo: image.stream.h264[0].masterUrl / url
      let livePhotoVideoUrl: string | null = null;
      const stream = image['stream'];
      if (isObject(stream)) {
        const h264 = stream['h264'];
        if (isArray(h264) && h264.length > 0) {
          const first = h264[0];
          if (isObject(first)) {
            livePhotoVideoUrl = asString(first['masterUrl']) ?? asString(first['url']);
          }
        }
      }

      if (imageUrl) {
        if (livePhotoVideoUrl) {
          mediaPairs.push({
            originalImageUrl: imageUrl,
            originalVideoUrl: livePhotoVideoUrl,
            imageUrl,
            videoUrl: livePhotoVideoUrl,
            isLivePhoto: true,
          });
          videosDetected = true;
          if (onVideoDetected) onVideoDetected();
        } else {
          mediaPairs.push({
            originalImageUrl: imageUrl,
            originalVideoUrl: null,
            imageUrl,
            videoUrl: null,
            isLivePhoto: false,
          });
        }
        // 修复 Java 版的 bug:图片 URL 也要进返回值
        mediaUrls.push(imageUrl);
      }
    }
  }

  return mediaUrls;
}

// 抑制 unused warning: videosDetected 在 Kotlin 内部用于 shouldCreateLivePhotos 决策,
// 这里仅作为副作用占位,TS 侧由上层判定。
export const _videosDetectedPlaceholder = (b: boolean) => b;
